python_modules/dagster/dagster/_utils/caching_instance_queryer.py

```diff
@@ -572,82 +410,131 @@ def has_dynamic_partition(self, partitions_def_name: str, partition_key: str) ->
     # RECONCILIATION
     ####################
 
-    def get_latest_storage_id(self, event_type: DagsterEventType) -> Optional[int]:
-        """Returns the latest storage id for an event of the given event type. If no such event
-        exists, returns None.
-
-        Args:
-            event_type (DagsterEventType): The event type to search for.
-        """
-        from dagster._core.event_api import EventRecordsFilter
-
-        records = list(
-            self.instance.get_event_records(
-                event_records_filter=EventRecordsFilter(event_type=event_type), limit=1
+    def _asset_partitions_data_versions(
+        self,
+        asset_key: AssetKey,
+        asset_partitions: Optional[Sequence[AssetKeyPartitionKey]],
+        after_cursor: Optional[int] = None,
+        before_cursor: Optional[int] = None,
+    ) -> Mapping[AssetKeyPartitionKey, Optional[DataVersion]]:
+        if not self.asset_graph.is_partitioned(asset_key):
+            asset_partition = AssetKeyPartitionKey(asset_key)
+            latest_record = self.get_latest_record(
+                asset_partition, after_cursor=after_cursor, before_cursor=before_cursor
+            )
+            return (
+                {asset_partition: extract_data_version_from_entry(latest_record.event_log_entry)}
+                if latest_record is not None
+                else {}
             )
-        )
-        if records:
-            return records[0].storage_id
         else:
-            return None
+            query_result = self.instance._event_storage.get_latest_tags_by_partition(  # noqa
+                asset_key,
+                event_type=self._event_type_for_key(asset_key),
+                tag_keys=[DATA_VERSION_TAG],
+                after_cursor=after_cursor,
+                before_cursor=before_cursor,
+                asset_partitions=[
+                    asset_partition.partition_key
+                    for asset_partition in asset_partitions
+                    if asset_partition.partition_key is not None
+                ]
+                if asset_partitions is not None
+                else None,
+            )
+            return {
+                AssetKeyPartitionKey(asset_key, partition_key): DataVersion(tags[DATA_VERSION_TAG])
+                if tags.get(DATA_VERSION_TAG)
+                else None
+                for partition_key, tags in query_result.items()
+            }
+
+    def get_new_asset_partitions(
```

Comment:
This method name makes me expect that it's returning asset partitions that previously didn't exist, independently of what has been materialized. I think it should have a more specific name, as well as include docstring that explains what precisely it's aiming to return.